File: data structures and algorithms/python/08_046.py
class TreeNode:
  def __init__(self, val=0, left=None, right=None, 
              connection=None):
     self.val = val
     self.left = left
     self.right = right
     self.connection = connection

def cloneTree(root):
  if not root:
     return None
  
  # Phase 1: Clone tree structure and create mapping
  node_map = {}
  
  def cloneStructure(node):
     if not node:
        return None
     
     if node in node_map:
        return node_map[node]
     
     new_node = TreeNode(node.val)
     node_map[node] = new_node
     
     new_node.left = cloneStructure(node.left)
     new_node.right = cloneStructure(node.right)
     
     return new_node
  
  new_root = cloneStructure(root)
  
  # Phase 2: Set connection pointers
  def setConnections(node):
     if not node:
        return
     
     if node.connection:
        node_map[node].connection = node_map[node.connection]
     
     setConnections(node.left)
     setConnections(node.right)
  
  setConnections(root)
  
  return new_root

File: data structures and algorithms/cs/08_046.cs
using System.Collections.Generic;

public class TreeNode {
   public int Val { get; set; }
   public TreeNode Left { get; set; }
   public TreeNode Right { get; set; }
   public TreeNode Connection { get; set; }

   public TreeNode(int val) {
      Val = val;
   }
}

public class Solution {
   public TreeNode CloneTree(TreeNode root) {
      if (root == null) {
         return null;
      }

      var nodeMap = new Dictionary<TreeNode, TreeNode>();
      var newRoot = CloneStructure(root, nodeMap);
      SetConnections(root, nodeMap);
      return newRoot;
   }

   private TreeNode CloneStructure(
      TreeNode node, 
      Dictionary<TreeNode, TreeNode> nodeMap
   ) {
      if (node == null) {
         return null;
      }

      if (nodeMap.ContainsKey(node)) {
         return nodeMap[node];
      }

      var newNode = new TreeNode(node.Val);
      nodeMap[node] = newNode;
      newNode.Left = CloneStructure(node.Left, nodeMap);
      newNode.Right = CloneStructure(node.Right, nodeMap);
      return newNode;
   }

   private void SetConnections(
      TreeNode originalNode,
      Dictionary<TreeNode, TreeNode> nodeMap
   ) {
      if (originalNode == null) {
         return;
      }

      if (originalNode.Connection != null) {
         nodeMap[originalNode].Connection = 
            nodeMap[originalNode.Connection];
      }

      SetConnections(originalNode.Left, nodeMap);
      SetConnections(originalNode.Right, nodeMap);
   }
}